### PR TITLE
Add opt-in AutoCloneType for Records

### PR DESCRIPTION
**Pull request by jackkoenig**
There is a new trait, chisel3.experimental.AutoCloneType that is mixed in to Bundle and can optionally be mixed in to user-defined Records. The compiler plugin prints a deprecation warning on any user-defined implementation of cloneType, telling the user to mix in AutoCloneType before upgrading to 3.6.

This is intended to be backported to 3.5.x for release in 3.5.5. I have checked that this backports to 3.5.x [relatively] cleanly and is binary compatible (with the one caveat that MixedVec mixing in `AutoCloneType` changes the tyep of it's `cloneType` in a way that violates binary compatibility. It might be a false positive, but regardless I will revert that aspect of this PR on the backport).

I also intend to follow this PR with a version making cloneType generation mandatory for Records in 3.6.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

  - code refactoring  
  - new feature/API

#### API Impact

This makes it possible for users to opt-in to autoclonetype for Records. Implementing `cloneType` for Records will be deprecated in 3.5.5 and an error in 3.6. This PR does not yet change the API but the API change is coming.

#### Backend Code Generation Impact

No impact

#### Desired Merge Strategy

 - Squash

#### Release Notes

Implement autoclonetype for `Records` with new `chisel3.experimental.AutoCloneType`. Implementing `cloneType` yourself is now a warning and will become an error in Chisel 3.6.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by jackkoenig** [comment]
Note that the changes to the BundleComponent look larger than they are, I factored out some of the logic into methods because the logic was long and hard to follow (and now needed to support `Records`).

**Review by chick — APPROVED** [review]
I think this looks fine, I looked at it pretty well and did some searches to try and see the re-use despite diff's weaknesses.
Looks ready to me
